modules/repository-datastax/src/main/java/com/intuit/wasabi/repository/cassandra/impl/AssignmentStats.java

```diff
@@ -0,0 +1,112 @@
+    private static Map<Integer, Map<String, AtomicInteger>> hourlyCountMap;
+    private static final Object lock = new Object();
+    private final int UUID_LENGTH = 36;
+
+    /**
+     * Constructor
+     */
+    @Inject
+    public AssignmentStats(HourlyBucketCountAccessor hourlyBucketCountAccessor) {
+        this.hourlyBucketCountAccessor = hourlyBucketCountAccessor;
+        hourlyCountMap = new ConcurrentHashMap<>();
+        for (int hour = 0; hour <= 23; hour++) {
+            hourlyCountMap.put(hour, new ConcurrentHashMap<>());
+        }
+    }
+
+    /**
+     * Increments the count of a particular bucket for a specific hour
+     *
+     * @param experiment
+     * @param assignment
+     */
+    public void incrementCount(Experiment experiment, Assignment assignment) {
+        int assignmentHour = AssignmentStatsUtil.getHour(new Date(System.currentTimeMillis()));
+        Map<String, AtomicInteger> hourMap = hourlyCountMap.get(assignmentHour);
+        Experiment.ID id = experiment.getID();
+        Optional<Bucket.Label> labelOptional = Optional.ofNullable(assignment.getBucketLabel());
+        Bucket.Label bucketLabel = labelOptional.orElseGet(() -> NULL_LABEL);
+        String key = new ExpBucket(id, bucketLabel).getKey();
+        AtomicInteger oldCount = hourMap.get(key);
+        if (oldCount == null) {
+            synchronized (lock) {                         // oldCount would be initialized to 1 twice w/o this
+                oldCount = hourMap.get(key);
+                if (oldCount == null) {                   // double-checked locking
+                    AtomicInteger count = new AtomicInteger(1);
+                    hourMap.put(key, count);
+                } else {
+                    oldCount.getAndIncrement();
+                }
+            }
+        } else {
+            oldCount.getAndIncrement();
+        }
+    }
+
+    /**
+     * Returns the number of assignments for an experiment and bucket during a specified hour
+     *
+     * @param experiment
+     * @param bucketLabel
+     * @param assignmentHour
+     * @return int representing the counts for a given bucket during the given hour
+     */
+    public int getCount(Experiment experiment, Bucket.Label bucketLabel, int assignmentHour) {
+        Map<String, AtomicInteger> hourMap = hourlyCountMap.get(assignmentHour);
+        return hourMap.get(new ExpBucket(experiment.getID(), bucketLabel).getKey()).get();
+    }
+
+    /**
+     * Writes hourly assignment counts to cassandra for the last completed hour based on current time
+     */
+    public void writeCounts() {
+        Date completedHour = AssignmentStatsUtil.getLastCompletedHour(System.currentTimeMillis());
+        int assignmentHour = AssignmentStatsUtil.getHour(completedHour);
+        String day = AssignmentStatsUtil.getDayString(completedHour);
+
+        for (String key : hourlyCountMap.get(assignmentHour).keySet()){
+            String experimentID = key.substring(0, UUID_LENGTH);
```

Comment:
@mooxter it would be better to use 'ExpBucket' as a key for second level map so that you dont have to deal with the parsing logic here... E.g. Map<Integer, Map<ExpBucket, AtomicInteger>> hourlyCountMap - instead of - Map<Integer, Map<String, AtomicInteger>> hourlyCountMap